### PR TITLE
inductor: fix customer op _convolution_pointwise_.binary functional error at AOTAutograd

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94581

This is another try(first is https://github.com/pytorch/pytorch/pull/94172) to fix the warning message when running inductor CPU path:

```
l.  Known situations this can occur are inference mode only compilation involving resize_ or prims (!schema.hasAnyAliasInfo() INTERNAL ASSERT FAILED); if your situation looks different please file a bug to PyTorch.
Traceback (most recent call last):
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 1377, in aot_wrapper_dedupe
    fw_metadata, _out = run_functionalized_fw_and_collect_metadata(flat_fn)(
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 578, in inner
    flat_f_outs = f(*flat_f_args)
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 2455, in functional_call
    out = Interpreter(mod).run(*args[params_len:], **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 136, in run
    self.env[node] = self.run_node(node)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 177, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 294, in call_module
    return submod(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/mkldnn.py", line 344, in forward
    return self._conv_forward(input, other, self.weight, self.bias)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/mkldnn.py", line 327, in _conv_forward
    return torch.ops.mkldnn._convolution_pointwise_(
  File "/home/xiaobing/pytorch-offical/torch/_ops.py", line 499, in __call__
    return self._op(*args, **kwargs or {})
  File "/home/xiaobing/pytorch-offical/torch/_inductor/overrides.py", line 38, in __torch_function__
    return func(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_ops.py", line 499, in __call__
    return self._op(*args, **kwargs or {})
RuntimeError: !schema.hasAnyAliasInfo() INTERNAL ASSERT FAILED at "/home/xiaobing/pytorch-offical/aten/src/ATen/FunctionalizeFallbackKernel.cpp":32, please report a bug to PyTorch. mutating and aliasing ops should all have codegen'd kernels

While executing %self_layer2_0_downsample_0 : [#users=2] = call_module[target=self_layer2_0_downsample_0](args = (%self_layer1_1_conv2, %self_layer2_0_conv2), kwargs = {})
Original traceback:
  File "/home/xiaobing/vision/torchvision/models/resnet.py", line 100, in forward
    identity = self.downsample(x)
 |   File "/home/xiaobing/vision/torchvision/models/resnet.py", line 274, in _forward_impl
    x = self.layer2(x)
 |   File "/home/xiaobing/vision/torchvision/models/resnet.py", line 285, in forward
    return self._forward_impl(x)
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/compile_fx.py [hunk 1]
```diff
@@ -398,6 +399,11 @@ def compile_fx(
     @dynamo_utils.dynamo_timed
     def fw_compiler(model: torch.fx.GraphModule, example_inputs):
         fixed = len(example_inputs) - num_example_inputs
+        # Why convert outplace op to inplace? For customer inplace op which for ExternalCall,
+        # there doesn't have a functionalization pass which will be failed when run the functionalization
+        # AOTAutograd step, we do this conversion for a good performance.
+        # We can safely do it which this only happen on inference mode.
```
Comment (jgong5) [inline torch/_inductor/compile_fx.py:405]:
```suggestion
        # Why convert outplace op to inplace? Inductor can support inplace operations well and for custom
        # inplace ops which are lowered as ExternKernel, it is beneficial to performance when the inplace
        # implementation is used if available.
```

**Review by jgong5 — APPROVED** [review]
LGTM. I added some comments/suggestions for code comments.

File: torch/_inductor/mkldnn.py [hunk 2]
```diff
@@ -789,47 +685,29 @@ def fuse_binary(gm: torch.fx.GraphModule):
...
-                    )
-                    # Make sure the fused node is post node of node's inputs nodes.
-                    node.append(node.args[1])
-                    gm.graph.erase_node(node)
-                    break
+        if node.op == "call_function" and node.target in [
+            torch.ops.mkldnn._convolution_pointwise.binary
+        ]:
+            # args[0] and args[2] is _convolution_pointwise.binary's input,
+            # need to check whether args[2] can be written or not.
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/_inductor/mkldnn.py:698]: Yes, changed.

File: torch/_inductor/mkldnn.py [hunk 3]
```diff
@@ -789,47 +685,29 @@ def fuse_binary(gm: torch.fx.GraphModule):
...
+            # TODO: node.args[1].users > 1, but node.args[1] never be used after current node.
+            if len(node.args[1].users) > 1:
+                continue
+            if node.args[1] == node.args[0]:
+                continue
+            binary_attr = node.args[8]
+            unary_attr = node.args[10]
+            if binary_attr != "add" or unary_attr not in ["", "relu"]:
+                continue
+            node.target = torch.ops.mkldnn._convolution_pointwise_.binary
```
Comment (jansel) [inline torch/_inductor/mkldnn.py:710]: Does the IR node for this op properly mark the mutation?

I worry inductor will try to reuse input buffer (which is now alised by the return value).

Comment (jgong5) [reply torch/_inductor/mkldnn.py:710]: Yes, it is marked mutation on creation: https://github.com/pytorch/pytorch/blob/master/torch/_inductor/ir.py#L3659

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**